File: yaf-gadomain/src/main/java/ar/edu/ungs/yamiko/ga/operators/impl/ProbabilisticRouletteSelector.java
package ar.edu.ungs.yamiko.ga.operators.impl;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

import ar.edu.ungs.yamiko.ga.domain.Individual;
import ar.edu.ungs.yamiko.ga.domain.Population;
import ar.edu.ungs.yamiko.ga.operators.Selector;
import ar.edu.ungs.yamiko.ga.toolkit.SearchHelper;
import ar.edu.ungs.yamiko.ga.toolkit.StaticHelper;

/**
 * Implementación de un selector por ruletas de probabilidades que trabaja sobre una población definida.
 * @author ricardo
 *
 */
public class ProbabilisticRouletteSelector<T> implements Selector<T>,Serializable{

	/**
	 * 
	 */
	private static final long serialVersionUID = -1021721399295562069L;

	public ProbabilisticRouletteSelector() {
		
	}
	
	@Override
	public Individual<T> execute(Population<T> p) {
		double roulette[]=buildRoulette(p);
		double rou=StaticHelper.randomDouble(roulette[roulette.length-1]);
		int selected=SearchHelper.binaryRangeSearch(roulette, rou);
		return (Individual<T>)p.getAll().get(selected);
	}	
	
	@Override
	public List<Individual<T>> executeN(int n,Population<T> p) {
		double roulette[]=buildRoulette(p);
		List<Individual<T>> salida=new ArrayList<Individual<T>>();
		for (int i=0;i<n;i++)
		{
			double rou=StaticHelper.randomDouble(roulette[roulette.length-1]);
			int selected=SearchHelper.binaryRangeSearch(roulette, rou);
			salida.add((Individual<T>)p.getAll().get(selected));			
		}
		return salida;
	}
	
	private double[] buildRoulette(Population<T> p)
	{
		double[] salida=new double[p.getAll().size()];
		double suma=0;
		for (int i=0;i<p.getAll().size();i++)
		{
			suma+=((Individual<T>)(p.getAll().get(i))).getFitness();
			salida[i]=suma;
		}
		return salida;
		
	}
}

File: yaf-gadomain-scala/src/main/scala/ar/edu/ungs/yamiko/ga/operators/impl/ProbabilisticRouletteSelector.scala
package ar.edu.ungs.yamiko.ga.operators.impl

import ar.edu.ungs.yamiko.ga.domain.Individual
import ar.edu.ungs.yamiko.ga.operators.Selector
import ar.edu.ungs.yamiko.ga.domain.Population
import scala.util.Random

/**
 * Implementación de un selector por ruletas de probabilidades que trabaja sobre una población definida.
 */
@SerialVersionUID(129L)
class ProbabilisticRouletteSelector[T] extends Selector[T] {
 
 private var k=10000
 
 override def getIntParameter=k
 override def setIntParameter(par:Int)={k=par}
 val DEFAULT_FACTOR=k.toDouble
 override def execute(p:Population[T]):Individual[T]=null
 
 override def executeN(n:Int,p:Population[T]):List[Individual[T]] =
 {
   val iNil:Individual[T]=null
   
   val div:Double=if (p.getAll()(0).getFitness==0d) DEFAULT_FACTOR else p.getAll()(0).getFitness()
   val accum=p.getAll.scanLeft((0d,0d,iNil)) { (a, i) => (a._1+ i.getFitness()/div,i.getFitness()/div,i) }.drop(1)
   val r=Random
   val max=accum.takeRight(1)(0)._1
   val salida=(1 to n).par.map { x => 
     val dRandom=r.nextDouble()*max 
     val sal=accum.find{s=>(s._1-s._2)<=dRandom && s._1>=dRandom}
     sal.get._3
     }.toList
	 return salida;   
 }
 
}